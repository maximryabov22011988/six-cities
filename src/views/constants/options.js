export const sortingName = {
  POPULAR: 'Popular',
  PRICE_LOW_TO_HIGH: 'Price: low to high',
  PRICE_HIGH_TO_LOW: 'Price: high to low',
  TOP_RATED: 'Top rated first',
};

export const SORTING_OPTIONS = [
  { id: 1, name: sortingName.POPULAR },
  { id: 2, name: sortingName.PRICE_LOW_TO_HIGH },
  { id: 3, name: sortingName.PRICE_HIGH_TO_LOW },
  { id: 4, name: sortingName.TOP_RATED },
];
