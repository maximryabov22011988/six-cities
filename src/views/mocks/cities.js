export default [
  {
    id: 1,
    city: 'Paris',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    id: 2,
    city: 'Cologne',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    id: 3,
    city: 'Brussels',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    id: 4,
    city: 'Amsterdam',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    id: 5,
    city: 'Hamburg',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    id: 6,
    city: 'Dusseldorf',
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  }
];
