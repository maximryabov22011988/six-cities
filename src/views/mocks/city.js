export default {
  name: 'Cologne',
  location: [50.938361, 6.959974],
  zoom: 13,
};
