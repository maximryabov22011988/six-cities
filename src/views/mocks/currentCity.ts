export default {
  name: 'Cologne',
  location: {
    latitude: 50.938361,
    longitude: 6.959974,
    zoom: 13,
  },
};
