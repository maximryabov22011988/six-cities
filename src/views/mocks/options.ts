const defaultId = 1;

const options = [{ id: 1, name: 'Popular' }];

export { defaultId, options };
