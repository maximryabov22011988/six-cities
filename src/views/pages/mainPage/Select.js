import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import withActiveItem from '../../hocs/withActiveItem';

import { ENTER, ESC, MOUSE_LEFT_BUTTON } from '../../constants/keyCodes';

const propTypes = {
  caption: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string
    })
  ).isRequired,
  defaultOption: PropTypes.number.isRequired
};

class Select extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentOption: props.defaultOption,
      isOpen: false
    };

    this.selectRef = React.createRef();
  }

  componentDidMount() {
    window.addEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  get currentOptionText() {
    const { currentOption } = this.state;
    const { options } = this.props;
    return options.find(item => item.id === currentOption).name;
  }

  handleOpen = () => {
    this.setState({
      isOpen: true
    });
  };

  handleClose = () => {
    this.setState({
      isOpen: false
    });
  };

  handleOutsideClick = evt => {
    evt.preventDefault();
    const { target } = evt;
    const { isOpen } = this.state;
    if (isOpen && !this.selectRef.current.contains(target)) {
      this.handleClose();
    }
  };

  handleOptionClick = id => evt => {
    const { nativeEvent } = evt;
    if (
      nativeEvent.which === MOUSE_LEFT_BUTTON ||
      nativeEvent.which === ENTER
    ) {
      this.setState({
        currentOption: id
      });
      this.handleClose();
    }
  };

  handleEscPress = evt => {
    const { nativeEvent } = evt;
    if (nativeEvent && nativeEvent.which === ESC) {
      this.handleClose();
    }
  };

  renderOptionItems = (options, currentOption) => {
    return options.map(({ id, name }) => {
      const Option = ({ className, onClick, onKeyUp, children }) => (
        <li
          className={className}
          tabIndex="0"
          onClick={onClick}
          onKeyUp={onKeyUp}
        >
          {children}
        </li>
      );
      const WrappedOption = withActiveItem(Option);

      return (
        <WrappedOption
          key={id}
          className="places__option"
          isActive={currentOption === id}
          onClick={this.handleOptionClick(id)}
          onKeyUp={this.handleOptionClick(id)}
        >
          {name}
        </WrappedOption>
      );
    });
  };

  render() {
    const { currentOption, isOpen } = this.state;
    const { caption, options } = this.props;

    return (
      <div
        className="places__sorting"
        tabIndex="0"
        ref={this.selectRef}
        onClick={this.handleOutsideClick}
        onKeyDown={this.handleEscPress}
      >
        {caption && <span className="places__sorting-caption">{caption}</span>}

        <span
          className="places__sorting-type"
          tabIndex="0"
          onClick={this.handleOpen}
          onFocus={this.handleOpen}
        >
          {this.currentOptionText}
          <svg className="places__sorting-arrow" width="7" height="4">
            <use xlinkHref="#icon-arrow-select" />
          </svg>
        </span>

        {isOpen && (
          <ul
            className={classnames(
              'places__options places__options--custom',
              isOpen && 'places__options--opened'
            )}
          >
            {this.renderOptionItems(options, currentOption)}
          </ul>
        )}
      </div>
    );
  }
}

Select.propTypes = propTypes;

export default Select;
