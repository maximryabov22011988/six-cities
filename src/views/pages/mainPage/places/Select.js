import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import withActiveItem from '../../../hocs/withActiveItem';

import { ENTER, ESC, MOUSE_LEFT_BUTTON } from '../../../constants/keyCodes';

const propTypes = {
  caption: PropTypes.string.isRequired,
  defaultOption: PropTypes.number.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
    })
  ).isRequired,
  onChangeSorting: PropTypes.func.isRequired,
};

class Select extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentOption: props.defaultOption,
      isOpen: false,
    };

    this.selectRef = React.createRef();
  }

  componentDidMount() {
    window.addEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  getCurrentOptionText() {
    const { currentOption } = this.state;
    const { options } = this.props;
    return options.find((item) => item.id === currentOption).name;
  }

  handleOpen = () => {
    this.setState({
      isOpen: true,
    });
  };

  handleClose = () => {
    this.setState({
      isOpen: false,
    });
  };

  handleOutsideClick = (evt) => {
    evt.preventDefault();
    const { target } = evt;
    const { isOpen } = this.state;
    if (isOpen && !this.selectRef.current.contains(target)) {
      this.handleClose();
    }
  };

  handleOptionClick = (id) => (evt) => {
    const { nativeEvent } = evt;
    const { onChangeSorting } = this.props;
    if (nativeEvent.which === MOUSE_LEFT_BUTTON || nativeEvent.which === ENTER) {
      this.setState(
        {
          currentOption: id,
        },
        () => onChangeSorting(this.getCurrentOptionText())
      );
      this.handleClose();
    }
  };

  handleEscPress = (evt) => {
    const { nativeEvent } = evt;
    if (nativeEvent && nativeEvent.which === ESC) {
      this.handleClose();
    }
  };

  renderOptionItems = (options, currentOption) => {
    return options.map(({ id, name }) => {
      const Option = ({ className, onClick, onKeyUp, children }) => (
        <li className={className} tabIndex="0" onClick={onClick} onKeyUp={onKeyUp}>
          {children}
        </li>
      );
      const WrappedOption = withActiveItem(Option);

      return (
        <WrappedOption
          className="places__option"
          isActive={currentOption === id}
          key={id}
          onClick={this.handleOptionClick(id)}
          onKeyUp={this.handleOptionClick(id)}
        >
          {name}
        </WrappedOption>
      );
    });
  };

  render() {
    const { currentOption, isOpen } = this.state;
    const { caption, options } = this.props;

    return (
      <div
        className="places__sorting"
        ref={this.selectRef}
        tabIndex="0"
        onClick={this.handleOutsideClick}
        onKeyDown={this.handleEscPress}
      >
        {caption && <span className="places__sorting-caption">{caption}</span>}

        <span className="places__sorting-type" tabIndex="0" onClick={this.handleOpen} onFocus={this.handleOpen}>
          {this.getCurrentOptionText()}
          <svg className="places__sorting-arrow" height="4" width="7">
            <use xlinkHref="#icon-arrow-select" />
          </svg>
        </span>

        {isOpen && (
          <ul className={cn('places__options places__options--custom', isOpen && 'places__options--opened')}>
            {this.renderOptionItems(options, currentOption)}
          </ul>
        )}
      </div>
    );
  }
}

Select.propTypes = propTypes;

export default Select;
