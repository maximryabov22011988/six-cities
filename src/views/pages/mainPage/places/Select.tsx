import * as React from 'react';
import cn from 'classnames';

import withActiveItem from '../../../hocs/withActiveItem';

import { ENTER, ESC, MOUSE_LEFT_BUTTON } from '../../../constants/keyCodes';

import { Option } from '../../../interfaces';

interface Props {
  caption: string,
  defaultOption: number,
  options: Array<Option>,
  onChangeSorting: (sorting: string) => void,
}

interface State {
  currentOption: number,
  isOpen: boolean,
}

class Select extends React.Component<Props, State> {
  selectRef = React.createRef<HTMLDivElement>();

  constructor(props) {
    super(props);
    this.state = {
      currentOption: props.defaultOption,
      isOpen: false,
    };
  }

  componentDidMount() {
    window.addEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.handleOutsideClick);
    window.addEventListener('keydown', this.handleEscPress);
  }

  getCurrentOptionText() {
    const { currentOption } = this.state;
    const { options } = this.props;
    return options.find((item) => item.id === currentOption).name;
  }

  handleOpen = () => {
    this.setState({
      isOpen: true,
    });
  };

  handleClose = () => {
    this.setState({
      isOpen: false,
    });
  };

  handleOutsideClick = (evt) => {
    evt.preventDefault();
    const { target } = evt;
    const { isOpen } = this.state;
    if (isOpen && !this.selectRef.current.contains(target)) {
      this.handleClose();
    }
  };

  handleOptionClick = (id) => (evt) => {
    const { nativeEvent } = evt;
    const { onChangeSorting } = this.props;
    if (nativeEvent.which === MOUSE_LEFT_BUTTON || nativeEvent.which === ENTER) {
      this.setState(
        {
          currentOption: id,
        },
        () => onChangeSorting(this.getCurrentOptionText())
      );
      this.handleClose();
    }
  };

  handleEscPress = (evt) => {
    const { nativeEvent } = evt;
    if (nativeEvent && nativeEvent.which === ESC) {
      this.handleClose();
    }
  };

  renderOptionItems = (options, currentOption) => {
    return options.map(({ id, name }) => {
      const Option = ({ className, onClick, onKeyUp, children }) => (
        <li className={className} tabIndex={0} onClick={onClick} onKeyUp={onKeyUp}>
          {children}
        </li>
      );
      const WrappedOption = withActiveItem(Option);

      return (
        <WrappedOption
          className="places__option"
          isActive={currentOption === id}
          key={id}
          onClick={this.handleOptionClick(id)}
          onKeyUp={this.handleOptionClick(id)}
        >
          {name}
        </WrappedOption>
      );
    });
  };

  render() {
    const { currentOption, isOpen } = this.state;
    const { caption, options } = this.props;

    return (
      <div
        className="places__sorting"
        ref={this.selectRef}
        tabIndex={0}
        onClick={this.handleOutsideClick}
        onKeyDown={this.handleEscPress}
      >
        {caption && <span className="places__sorting-caption">{caption}</span>}

        <span className="places__sorting-type" tabIndex={0} onClick={this.handleOpen} onFocus={this.handleOpen}>
          {this.getCurrentOptionText()}
          <svg className="places__sorting-arrow" height="4" width="7">
            <use xlinkHref="#icon-arrow-select" />
          </svg>
        </span>

        {isOpen && (
          <ul className={cn('places__options places__options--custom', isOpen && 'places__options--opened')}>
            {this.renderOptionItems(options, currentOption)}
          </ul>
        )}
      </div>
    );
  }
}

export default Select;
