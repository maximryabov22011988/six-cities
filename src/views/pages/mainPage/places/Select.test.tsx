import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Select from './Select';
import { defaultId, options } from '../../../mocks/options';

it('Select renders correctly with mandatory props', () => {
  const tree = renderer
    .create(
      <Select
        caption="Sort by"
        defaultOption={defaultId}
        options={options}
        onChangeSorting={jest.fn()}
      />
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
