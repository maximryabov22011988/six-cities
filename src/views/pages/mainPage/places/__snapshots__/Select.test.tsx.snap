// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly with mandatory props 1`] = `
<div
  className="places__sorting"
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    onFocus={[Function]}
    tabIndex={0}
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
</div>
`;
